File: todosv2/migrate-old-todos.js
const mysql = require('mysql2/promise');
const sqlite3 = require('sqlite3').verbose();
const bcrypt = require('bcrypt');
const path = require('path');

// Configuration
const OLD_DB_CONFIG = {
  host: '43.140.218.248',
  port: 3306,
  user: 'root',
  password: 'lidaning',
  database: 'todos',
  connectTimeout: 10000
};

const NEW_DB_PATH = path.join(__dirname, 'todos.db');

async function connectOldDb() {
  try {
    const connection = await mysql.createConnection(OLD_DB_CONFIG);
    console.log("✅ Connected to old MySQL database");
    return connection;
  } catch (error) {
    console.error("❌ Failed to connect to old database:", error.message);
    return null;
  }
}

function connectNewDb() {
  return new Promise((resolve, reject) => {
    const db = new sqlite3.Database(NEW_DB_PATH, sqlite3.OPEN_READWRITE, (err) => {
      if (err) {
        console.error("❌ Failed to connect to new database:", err.message);
        reject(err);
      } else {
        console.log("✅ Connected to new SQLite database");
        resolve(db);
      }
    });
  });
}

async function getOrCreateUser(db) {
  return new Promise((resolve, reject) => {
    // Check if test user exists
    db.get("SELECT id FROM users WHERE email = 'test@example.com'", async (err, user) => {
      if (err) {
        reject(err);
        return;
      }
      
      if (user) {
        console.log(`✅ Using existing user with ID: ${user.id}`);
        resolve(user.id);
        return;
      }
      
      // Create test user
      const hashedPassword = await bcrypt.hash('password123', 10);
      db.run(
        "INSERT INTO users (email, username, password) VALUES (?, ?, ?)",
        ['test@example.com', 'Test User', hashedPassword],
        function(err) {
          if (err) {
            reject(err);
            return;
          }
          const userId = this.lastID;
          console.log(`✅ Created new user with ID: ${userId}`);
          resolve(userId);
        }
      );
    });
  });
}

async function migrateCategories(oldConn, newDb, userId) {
  console.log("\n📁 Migrating categories...");
  
  try {
    const [oldCategories] = await oldConn.execute("SELECT id, catagory_name FROM catagory");
    
    const categoryMap = {};
    let createdCount = 0;
    let existingCount = 0;
    
    for (const oldCat of oldCategories) {
      const oldId = oldCat.id;
      const name = oldCat.catagory_name;
      
      // Check if category already exists
      const existing = await new Promise((resolve, reject) => {
        newDb.get(
          "SELECT id FROM categories WHERE name = ? AND userId = ?",
          [name, userId],
          (err, row) => {
            if (err) reject(err);
            else resolve(row);
          }
        );
      });
      
      if (existing) {
        categoryMap[oldId] = existing.id;
        existingCount++;
        console.log(`  ✓ Category '${name}' already exists (ID: ${existing.id})`);
      } else {
        // Create new category with default color and icon
        const colors = ['#3B82F6', '#10B981', '#8B5CF6', '#EF4444', '#F59E0B'];
        const icons = ['💼', '👤', '🛒', '❤️', '📚'];
        
        const color = colors[createdCount % colors.length];
        const icon = icons[createdCount % icons.length];
        
        const newId = await new Promise((resolve, reject) => {
          newDb.run(
            "INSERT INTO categories (name, color, icon, userId) VALUES (?, ?, ?, ?)",
            [name, color, icon, userId],
            function(err) {
              if (err) reject(err);
              else resolve(this.lastID);
            }
          );
        });
        
        categoryMap[oldId] = newId;
        createdCount++;
        console.log(`  ✓ Created category '${name}' (ID: ${newId})`);
      }
    }
    
    console.log(`✅ Migrated ${oldCategories.length} categories (${createdCount} created, ${existingCount} existing)`);
    return categoryMap;
  } catch (error) {
    console.error("❌ Failed to migrate categories:", error.message);
    return {};
  }
}

async function migrateTodos(oldConn, newDb, userId, categoryMap) {
  console.log("\n✅ Migrating todos...");
  
  try {
    const [oldTodos] = await oldConn.execute(`
      SELECT id, user_id, catagory_id, catagory_name, todo_name, 
             acomplished_time, create_time, is_completed, postponed
      FROM todos
      ORDER BY id
    `);
    
    let migratedCount = 0;
    let skippedCount = 0;
    
    for (const oldTodo of oldTodos) {
      const oldId = oldTodo.id;
      const todoName = oldTodo.todo_name;
      const isCompleted = oldTodo.is_completed === '1';
      const postponedCount = oldTodo.postponed || 0;
      const oldCategoryId = oldTodo.catagory_id;
      
      // Check if todo already exists
      const existing = await new Promise((resolve, reject) => {
        newDb.get(
          "SELECT id FROM todos WHERE title = ? AND userId = ?",
          [todoName, userId],
          (err, row) => {
            if (err) reject(err);
            else resolve(row);
          }
        );
      });
      
      if (existing) {
        skippedCount++;
        console.log(`  ⏭️  Todo '${todoName.substring(0, 30)}...' already exists, skipping`);
        continue;
      }
      
      // Map priority based on postponed count
      let priority = 'low';
      if (postponedCount >= 3) {
        priority = 'high';
      } else if (postponedCount >= 1) {
        priority = 'medium';
      }
      
      // Map category
      let newCategoryId = null;
      if (oldCategoryId && categoryMap[oldCategoryId]) {
        newCategoryId = categoryMap[oldCategoryId];
      }
      
      // Convert dates
      const createdAt = oldTodo.create_time 
        ? new Date(oldTodo.create_time).toISOString().replace('T', ' ').substring(0, 19)
        : new Date().toISOString().replace('T', ' ').substring(0, 19);
      
      let completedAt = null;
      if (isCompleted && oldTodo.acomplished_time) {
        completedAt = new Date(oldTodo.acomplished_time).toISOString().replace('T', ' ').substring(0, 19);
      }
      
      // Insert into new database
      await new Promise((resolve, reject) => {
        newDb.run(`
          INSERT INTO todos (
            title, description, priority, dueDate, completedAt, 
            isCompleted, postponedCount, estimatedMinutes, tags, 
            userId, categoryId, createdAt
          ) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)
        `, [
          todoName,
          `Migrated from old todo ID: ${oldId}`,
          priority,
          null, // dueDate
          completedAt,
          isCompleted ? 1 : 0,
          postponedCount,
          30, // Default estimated minutes
          JSON.stringify(['migrated']),
          userId,
          newCategoryId,
          createdAt
        ], function(err) {
          if (err) reject(err);
          else resolve();
        });
      });
      
      migratedCount++;
      const status = isCompleted ? "✓" : "⏳";
      console.log(`  ${status} Migrated todo '${todoName.substring(0, 30)}...' (Priority: ${priority})`);
    }
    
    console.log(`✅ Migrated ${migratedCount} todos, skipped ${skippedCount} duplicates`);
    return migratedCount;
  } catch (error) {
    console.error("❌ Failed to migrate todos:", error.message);
    return 0;
  }
}

async function main() {
  console.log("🚀 Starting migration from old MySQL to new SQLite database");
  console.log("=" .repeat(60));
  
  let oldConn = null;
  let newDb = null;
  
  try {
    // Connect to databases
    oldConn = await connectOldDb();
    if (!oldConn) {
      throw new Error("Failed to connect to old database");
    }
    
    newDb = await connectNewDb();
    
    // Get or create user
    const userId = await getOrCreateUser(newDb);
    
    // Migrate categories
    const categoryMap = await migrateCategories(oldConn, newDb, userId);
    
    // Migrate todos
    const migratedCount = await migrateTodos(oldConn, newDb, userId, categoryMap);
    
    console.log("\n" + "=" .repeat(60));
    console.log("🎉 Migration completed successfully!");
    console.log("📊 Summary:");
    console.log(`   • User ID: ${userId}`);
    console.log(`   • Categories mapped: ${Object.keys(categoryMap).length}`);
    console.log(`   • Todos migrated: ${migratedCount}`);
    console.log("\n🔑 Login credentials:");
    console.log("   Email: test@example.com");
    console.log("   Password: password123");
    console.log("\n🌐 Access the app at: http://localhost:3000");
    
  } catch (error) {
    console.error("\n❌ Migration failed:", error.message);
  } finally {
    // Close connections
    if (oldConn) {
      await oldConn.end();
    }
    if (newDb) {
      newDb.close();
    }
    console.log("\n🔒 Database connections closed");
  }
}

// Run migration
main().catch(console.error);

File: todosv2/migrate-old-todos.py
#!/usr/bin/env python3
"""
Migration script to transfer todos from old MySQL database to new SQLite database
"""

import mysql.connector
import sqlite3
import json
from datetime import datetime
import bcrypt

# Configuration
OLD_DB_CONFIG = {
    'host': '43.140.218.248',
    'port': 3306,
    'user': 'root',
    'password': 'lidaning',
    'database': 'todos'
}

NEW_DB_PATH = './todos.db'

def connect_old_db():
    """Connect to the old MySQL database"""
    try:
        conn = mysql.connector.connect(**OLD_DB_CONFIG)
        print("✅ Connected to old MySQL database")
        return conn
    except Exception as e:
        print(f"❌ Failed to connect to old database: {e}")
        return None

def connect_new_db():
    """Connect to the new SQLite database"""
    try:
        conn = sqlite3.connect(NEW_DB_PATH)
        conn.row_factory = sqlite3.Row
        print("✅ Connected to new SQLite database")
        return conn
    except Exception as e:
        print(f"❌ Failed to connect to new database: {e}")
        return None

def get_or_create_user(new_cursor):
    """Get or create a user in the new database"""
    # Check if test user exists
    new_cursor.execute("SELECT id FROM users WHERE email = 'test@example.com'")
    user = new_cursor.fetchone()
    
    if user:
        print(f"✅ Using existing user with ID: {user['id']}")
        return user['id']
    
    # Create test user
    hashed_password = bcrypt.hashpw('password123'.encode('utf-8'), bcrypt.gensalt())
    new_cursor.execute(
        "INSERT INTO users (email, username, password) VALUES (?, ?, ?)",
        ('test@example.com', 'Test User', hashed_password.decode('utf-8'))
    )
    user_id = new_cursor.lastrowid
    print(f"✅ Created new user with ID: {user_id}")
    return user_id

def migrate_categories(old_cursor, new_cursor, user_id):
    """Migrate categories from old to new database"""
    print("\n📁 Migrating categories...")
    
    # Get categories from old database
    old_cursor.execute("SELECT id, catagory_name FROM catagory")
    old_categories = old_cursor.fetchall()
    
    category_map = {}  # Map old category ID to new category ID
    
    for old_cat in old_categories:
        old_id = old_cat[0]
        name = old_cat[1]
        
        # Check if category already exists
        new_cursor.execute(
            "SELECT id FROM categories WHERE name = ? AND userId = ?",
            (name, user_id)
        )
        existing = new_cursor.fetchone()
        
        if existing:
            category_map[old_id] = existing['id']
            print(f"  ✓ Category '{name}' already exists (ID: {existing['id']})")
        else:
            # Create new category with default color and icon
            colors = ['#3B82F6', '#10B981', '#8B5CF6', '#EF4444', '#F59E0B']
            icons = ['💼', '👤', '🛒', '❤️', '📚']
            
            color = colors[len(category_map) % len(colors)]
            icon = icons[len(category_map) % len(icons)]
            
            new_cursor.execute(
                "INSERT INTO categories (name, color, icon, userId) VALUES (?, ?, ?, ?)",
                (name, color, icon, user_id)
            )
            new_id = new_cursor.lastrowid
            category_map[old_id] = new_id
            print(f"  ✓ Created category '{name}' (ID: {new_id})")
    
    print(f"✅ Migrated {len(category_map)} categories")
    return category_map

def migrate_todos(old_cursor, new_cursor, user_id, category_map):
    """Migrate todos from old to new database"""
    print("\n✅ Migrating todos...")
    
    # Get todos from old database
    old_cursor.execute("""
        SELECT id, user_id, catagory_id, catagory_name, todo_name, 
               acomplished_time, create_time, is_completed, postponed
        FROM todos
        ORDER BY id
    """)
    old_todos = old_cursor.fetchall()
    
    migrated_count = 0
    skipped_count = 0
    
    for old_todo in old_todos:
        old_id = old_todo[0]
        old_user_id = old_todo[1]
        old_category_id = old_todo[2]
        old_category_name = old_todo[3]
        todo_name = old_todo[4]
        accomplished_time = old_todo[5]
        create_time = old_todo[6]
        is_completed = old_todo[7] == '1'
        postponed_count = old_todo[8]
        
        # Check if todo already exists (by title and user)
        new_cursor.execute(
            "SELECT id FROM todos WHERE title = ? AND userId = ?",
            (todo_name, user_id)
        )
        existing = new_cursor.fetchone()
        
        if existing:
            skipped_count += 1
            print(f"  ⏭️  Todo '{todo_name[:30]}...' already exists, skipping")
            continue
        
        # Map priority based on postponed count
        if postponed_count >= 3:
            priority = 'high'
        elif postponed_count >= 1:
            priority = 'medium'
        else:
            priority = 'low'
        
        # Map category
        new_category_id = None
        if old_category_id and old_category_id in category_map:
            new_category_id = category_map[old_category_id]
        
        # Convert dates
        created_at = None
        if create_time:
            created_at = create_time.strftime('%Y-%m-%d %H:%M:%S') if isinstance(create_time, datetime) else create_time
        
        completed_at = None
        if is_completed and accomplished_time:
            completed_at = accomplished_time.strftime('%Y-%m-%d %H:%M:%S') if isinstance(accomplished_time, datetime) else accomplished_time
        
        # Insert into new database
        new_cursor.execute("""
            INSERT INTO todos (
                title, description, priority, dueDate, completedAt, 
                isCompleted, postponedCount, estimatedMinutes, tags, 
                userId, categoryId, createdAt
            ) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)
        """, (
            todo_name,
            f"Migrated from old todo ID: {old_id}",
            priority,
            None,  # dueDate
            completed_at,
            1 if is_completed else 0,
            int(postponed_count) if postponed_count else 0,
            30,  # Default estimated minutes
            json.dumps(['migrated']),
            user_id,
            new_category_id,
            created_at or datetime.now().strftime('%Y-%m-%d %H:%M:%S')
        ))
        
        migrated_count += 1
        status = "✓" if is_completed else "⏳"
        print(f"  {status} Migrated todo '{todo_name[:30]}...' (Priority: {priority})")
    
    print(f"✅ Migrated {migrated_count} todos, skipped {skipped_count} duplicates")
    return migrated_count

def main():
    print("🚀 Starting migration from old MySQL to new SQLite database")
    print("=" * 60)
    
    # Connect to databases
    old_conn = connect_old_db()
    if not old_conn:
        return
    
    new_conn = connect_new_db()
    if not new_conn:
        old_conn.close()
        return
    
    try:
        old_cursor = old_conn.cursor()
        new_cursor = new_conn.cursor()
        
        # Get or create user
        user_id = get_or_create_user(new_cursor)
        
        # Migrate categories
        category_map = migrate_categories(old_cursor, new_cursor, user_id)
        
        # Migrate todos
        migrated_count = migrate_todos(old_cursor, new_cursor, user_id, category_map)
        
        # Commit changes
        new_conn.commit()
        
        print("\n" + "=" * 60)
        print(f"🎉 Migration completed successfully!")
        print(f"📊 Summary:")
        print(f"   • User ID: {user_id}")
        print(f"   • Categories migrated: {len(category_map)}")
        print(f"   • Todos migrated: {migrated_count}")
        print(f"\n🔑 Login credentials:")
        print(f"   Email: test@example.com")
        print(f"   Password: password123")
        print(f"\n🌐 Access the app at: http://localhost:3000")
        
    except Exception as e:
        print(f"\n❌ Migration failed: {e}")
        new_conn.rollback()
    finally:
        old_cursor.close()
        new_cursor.close()
        old_conn.close()
        new_conn.close()
        print("\n🔒 Database connections closed")

if __name__ == "__main__":
    main()